aioredis/cluster/cluster.py

```diff
@@ -0,0 +1,615 @@
+    def fetch_cluster_info(self):
+        logger.info('Loading cluster info from {}...'.format(self._nodes))
+        tasks = [async_task(self._get_raw_cluster_info_from_node(node),
+                            loop=self._loop) for node in self._nodes]
+
+        try:
+            for task in asyncio.as_completed(tasks, loop=self._loop):
+                try:
+                    nodes_raw_response = yield from task
+                    self._cluster_manager = ClusterNodesManager.create(
+                        nodes_raw_response)
+                    logger.info('Cluster info loaded successfully: %s',
+                                nodes_raw_response)
+                    return
+                except (ReplyError, ProtocolError, ConnectionError) as exc:
+                    logger.warning(
+                        "Loading cluster info from a node failed with {}"
+                        .format(repr(exc))
+                    )
+        finally:
+            for task in tasks:
+                task.cancel()
+            # Wait until all tasks have closed their connection
+            yield from asyncio.gather(
+                *tasks, loop=self._loop, return_exceptions=True)
+
+        raise RedisClusterError(
+            "No cluster info could be loaded from any host")
+
+    @asyncio.coroutine
+    def initialize(self):
+        logger.info('Initializing cluster...')
+        self._moved_count = 0
+        yield from self.fetch_cluster_info()
+        logger.info('Initialized cluster.\n{}'.format(self._cluster_manager))
+
+    @asyncio.coroutine
+    def clear(self):
+        pass  # All connections are created on demand and destroyed afterwards.
+
+    @asyncio.coroutine
+    def create_connection(self, address):
+        conn = yield from create_redis(
+            address, db=self._db, encoding=self._encoding,
+            password=self._password, loop=self._loop, )
+        return conn
+
+    @asyncio.coroutine
+    def _execute_node(self, address, command, *args, **kwargs):
+        """Execute redis command and returns Future waiting for the answer.
+
+        :param command str
+        :param pool obj
+        Raises:
+        * TypeError if any of args can not be encoded as bytes.
+        * ReplyError on redis '-ERR' responses.
+        * ProtocolError when response can not be decoded meaning connection
+          is broken.
+        """
+        cmd = decode(command).lower()
+        to_close = []
+        try:
+            conn = yield from self.create_connection(address)
+            to_close.append(conn)
+            return (yield from getattr(conn, cmd)(*args, **kwargs))
+        except ReplyError as err:
+            address = parse_moved_response_error(err)
+            if address is None:
+                raise
+            logger.debug('Got MOVED command: {}'.format(err))
+            self._moved_count += 1
+            if self._moved_count >= self.MAX_MOVED_COUNT:
+                yield from self.initialize()
```

Comment:
Without any locks you'll end up with the following:
* assume `_moved_count == MAX_MOVED_COUNT - 1`;
* two coroutines execute command on moved keys;
* first one increments `_moved_count`, it hits limit and `initialize()` is called, execution is passed to loop;
* loop resumes the second coro;
* it recieves moved error as well; increments `_moved_count` and calls `initialize()` too.
The cluster would be re-configured twice.
now imagine that there are more that 2 coros hits this case...